Fix for issue #40985

stefanuhrig opened the pull request:
The canvas selection rectangle might not be an axis-parallel rectangle in the layer's CRS. Hence, it is not sufficient to transform the lower-left and upper-right corner of the canvas rectangle.

We interpolate the canvas rectangle, transform it to the layer CRS and check if it is still a nearly axis-parallel rectangle. If it is (or if we are in the single-click case), we just take its bounding box as filter rectangle in the feature request. Otherwise, we additionally use the transformed rectangle as postfilter geometry.

Fixes #40985

stefanuhrig committed:
Add test for issue #40985

stefanuhrig committed:
Fix issue #40985

stefanuhrig commented on src/gui/qgsmaptool.cpp:
```diff
@@ -66,7 +66,7 @@ QgsPointXY QgsMapTool::toMapCoordinates( const QgsMapLayer *layer, const QgsPoin
 QgsRectangle QgsMapTool::toLayerCoordinates( const QgsMapLayer *layer, const QgsRectangle &rect )
 {
-  return mCanvas->mapSettings().mapToLayerCoordinates( layer, rect );
+  return mCanvas->mapSettings().outputExtentToLayerExtent( layer, rect );
```
I was not sure whether I can just change this to `outputExtentToLayerExtent` or whether I should introduce a new method in `QgsMapTool` and call that method from `QgsMapToolIdentify::identifyVectorLayer`.

Comments on that would be appreciated.

roya0045 commented:
It's nice to see you guys help fix things that are not only related to HANA. 👍

stefanuhrig committed:
Update qgsmaptool.sip.in

uclaros commented:
I'm afraid with this solution you do not get more precise results. You actually just enlarge the rectangle to its bbox on the layer crs. On one hand you can guarantee that all items in your rectangle are actually identified but on the other hand you will get more false positives!
The proper approach would be to implement something similar to how the selection tool works. See https://github.com/qgis/QGIS/blob/67d57541eef0d48058b26c1915ae556c7706ad81/src/app/qgsmaptoolselectutils.cpp#L166

stefanuhrig commented:
Thanks for the feedback, @uclaros.

I agree that this solution might lead to false positives. If I understand you correctly, you are proposing to interpolate the query geometry, to transform it to the vector layer CRS and to use this geometry to query the vector layer.

I'm afraid that this won't work because the `QgsFeatureRequest` class does not allow a filter geometry, but a filter rectangle only. Well, we could set a `QgsExpression` filter with an `intersects`-function, but that would probably be a bad idea. If a vector layer does not support the filter function natively, it will return all features and they will then be post-filtered in QGIS.

However, the maptool already supports non-rectangle query geometries. In that case, it uses the bounding box of the query geometry to query the vector layer and postfilters the results with the query geometry transformed to the vector layer CRS.

We could do that independent of the query geometry's type:

- Interpolate the query geometry.
- Transform it to the vector layer CRS.
- Use the bounding box as filter rectangle in the vector layer query.
- Postfilter with the interpolated and transformed geometry.

This would avoid both false negatives and false positives (well, depends on the granularity of the interpolation, but will probably be good enough), but it would require some potentially expensive post-processing.

To summarize, I see the following options:

1. We leave it as it is and accept false negatives.
2. We use the current approach of this pull request and accept false positives.
3. We use the sophisticated interpolation approach and accept the performance impact.

What do you think? Any other ideas?

nyalldawson commented on src/gui/qgsmaptool.h:
```diff
@@ -278,7 +278,7 @@ class GUI_EXPORT QgsMapTool : public QObject
     QgsPoint toMapCoordinates( const QgsMapLayer *layer, const QgsPoint &point ) SIP_PYNAME( toMapCoordinatesV2 );
 
-    //! trnasformation of the rect from map coordinates to layer's coordinates
+    //! transformation of the rect from map coordinates to layer's coordinates
```
```suggestion
    //! Transforms a \a rect from map coordinates to a \a layer's coordinates.
```

May as well fix the whole string up :smile:

stefanuhrig replied:
There are more strings in the same style in the file. I'll change them in a separate pull request to the format you suggested.

stefanuhrig replied:
I've opened #41216 for the class member documentation improvements.

stefanuhrig commented:
I had another idea:

We could transform the query rectangle to the vector layer CRS (all four corners plus maybe one intermediate point per side) and then check if the resulting geometry is still roughly a rectangle.

If it is still roughly a rectangle, we take the bounding box of the transformed rectangle as filter rectangle and don't perform any postfiltering.

If it is not a rectangle anymore, we do a finer interpolation of the query rectangle, transform it to the vector layer, filter with the bounding box and postfilter with the transformed query rectangle.

Then we have no negative performance impact if the rectangle is still a rectangle after transformation (which is probably often the case). Otherwise, the postfilter assures that we have no false positives (comes at a cost though).

What do you think? Should we try that approach?

roya0045 commented:
I fear that doing many reprojection may be costly but it would be the best way to handle this.

stefanuhrig committed:
Merge branch 'master' of https://github.com/qgis/QGIS into issue40985

stefanuhrig commented:
Okay, I will give it a try and will prepare an update to this pull request.

stefanuhrig committed:
A more sophisticated approach for identifying features

The new approach attempts to avoid false negatives and false positives
while still trying to avoid unnecessary postfilters.

stefanuhrig committed:
Fix formatting

stefanuhrig committed:
Fix index-out-of-bounds-access

nyalldawson commented:
@stefanuhrig I think there's a simpler approach you could use here.

Instead of 

```
r = toLayerCoordinates( layer, selectionGeom.boundingBox() );
```

at line 596 of qgsmaptoolidentify.cpp (which uses `QgsCoordinateTransform::transform( const QgsRectangle &rect, TransformDirection direction ) const )`, we instead need to be using `QgsCoordinateTransform::transformBoundingBox` here.
QgsCoordinateTransform::transformBoundingBox already has similar logic to what you've added, and is guaranteed to return a rectangle which completely covers the original source rectangle (i.e. it expands the rectangle upon reprojection). (Side note: your logic looks much more intelligent than that in transformBoundingBox, so I'd suggest you submit your refinements as a seperate PR reworking the existing transformBoundingBox method, which is heavily used throughout QGIS and would benefit from your improvements!!)

So potentially this bug could be fixed by adding a new boolean argument to `QgsMapTool::toLayerCoordinates( const QgsMapLayer *layer, const QgsRectangle &rect )` and `QgsMapSettings::mapToLayerCoordinates( const QgsMapLayer *layer, QgsRectangle rect )` for something like "bool useBoundingBoxTransform", which would cause QgsMapSettings::mapToLayerCoordinates to use ct.transformBoundingBox(...) instead of just ct.transform(...).

What do you think?

stefanuhrig commented:
@nyalldawson Thanks for the feedback!

Originally, I implemented the approach you are suggesting. However, @uclaros argued (https://github.com/qgis/QGIS/pull/41172#issuecomment-767126443) that this approach might return false positives.

To avoid false positives, we need to postfilter the results returned by the vector layer with the transformed selection rectangle if it is not (almost) an axis-parallel rectangle.

That's why I added methods to create an approximation of the transformed selection rectangle and to check if it is almost an axis-parallel rectangle.

I'm not sure if the "almost-an-axis-parallel-rectangle"-check is the best way to do it. I also considered the discrete Hausdorff distance between the transformed rectangle and its bounding box, but was unsure how to parameterize it.

Anyway, as the "Identify feature" tool is vital, I'd suggest to postpone this fix to 3.20. That leaves us some more room for discussions, so that we hopefully won't break something in a haste.

nyalldawson commented:
> Originally, I implemented the approach you are suggesting. However, @uclaros argued (#41172 (comment)) that this approach might return false positives.

Hm - what about an alternative simple approach then and modifying the second pass filtering done between lines 617-620 so that it sets the destination crs on the feature request to the ORIGINAL crs for the selection geometry, and then we compare against the original selection geometry in the intersects tests instead of the transformed one?

stefanuhrig commented:
@nyalldawson That might be a good idea! It would be the most accurate approach, also with regards to non-rectangular selection geometries.

Let me structure my own thoughts on that first. When a line string or a polygon is transformed from one CRS to another one, usually only the vertices defining the geometry are transformed. After transformation, the vertices are connected by straight lines again. However, that's not accurate in many cases. If we transform all (intermediate) points on a straight line to another CRS, it's often not a straight line anymore afterwards. However, most of the time that's good enough and it's probably how it's done when geometries from a different CRS are rendered on the canvas.

Even if we interpolate the selection geometry with many intermediate points, transform that to the layer CRS, we cannot account for the inaccuracies introduced by the rendering. So, by requesting all possible candidates from the layer, transforming them to the canvas CRS and filtering with the selection geometry, the result would exactly match what a user sees on the canvas.

Drawback of that approach is that we have to transform all candidate geometries to the canvas CRS (instead of only one geometry to the layer CRS) and that we have to do the postfilter. However, both steps could be omitted if the selection rectangle would also be almost a rectangle in the layer CRS (so, we probably need an "almost-a-rectangle"-check).

Another advantage of that approach is that the postfiltering is cheaper. The current approach in this PR filters with the interpolated transformed rectangle, which currently has 20 edges. The proposed approach filters with the selection rectangle, which has only 4 edges. That might well over-compensate the costs of transforming all candidate geometries.

I'll prepare a new commit.
